src/gt4py/next/embedded/nd_array_field.py

```diff
@@ -174,12 +189,57 @@ def from_array(
     def remap(self: _BaseNdArrayField, connectivity) -> _BaseNdArrayField:
         raise NotImplementedError()
 
+    def __getitem__(self, index: common.FieldSlice) -> common.Field | core_defs.ScalarT:
+        index = _tuplize_field_slice(index)
+
+        if common.is_domain_slice(index):
+            return self._getitem_absolute_slice(index)
+
+        assert isinstance(index, tuple)
+        if all(
+            isinstance(idx, slice) or common.is_int_index(idx) or idx is Ellipsis for idx in index
+        ):
+            return self._getitem_relative_slice(index)
+
+        raise IndexError(f"Unsupported index type: {index}")
+
+    restrict = (
+        __getitem__  # type:ignore[assignment] # TODO(havogt) I don't see the problem that mypy has
+    )
+
+    def __setitem__(
+        self,
+        index: common.FieldSlice,
+        value: common.Field | core_defs.NDArrayObject | core_defs.ScalarT,
+    ) -> None:
+        index = _tuplize_field_slice(index)
+        if common.is_field(value):
+            # TODO(havogt): in case of `is_field(value)` we should additionally check that `value.domain == self[slice].domain`
+            value = value.ndarray
```

Comment:
I would stick to the usual pattern of putting the body of the function in `restrict` and the defining `__getitem__` as `restrict`